README.md
# krita-stable-diffusion

This repository includes a Krita plugin for communication with [stablehorde](https://stablehorde.net). Stablehorde is a cluster of stable-diffusion servers run by volunteers. You can create stable-diffusion images for free without running a colab notebook or a local server. Please check the section "Limitations" to better understand where the limits are.

Please check HISTORY.md for the latest changes. 

https://user-images.githubusercontent.com/113246030/196197783-ee750552-b4f9-4722-8a23-0116a1674c9e.MOV


## Installation
### Krita

The plugin has been tested in Krita 5.1.1.

**IMPORTANT:** If you are running Linux, please make sure, you have webp support for Qt5 installed, as this is the image format stablehorde uses. Execute the following command to check, if it's already supported: ```python -c "from PyQt5.Qt import *; print([bytes(x).decode('ascii') for x in QImageReader.supportedImageFormats()])"```. To install it, use ```sudo pacman -S qt5-imageformats``` (Arch) or ```sudo apt install qt5-image-formats-plugins``` (Debian).

1. Download the [Krita plugin zip](https://github.com/blueturtleai/krita-stable-diffusion/releases/download/v1.3.4/krita_stablehorde_1_3_4.zip).

2. Start Krita and open the "Tools/Scripts/Import Python Plugin from File" menu and select the downloaded zip. Restart Krita.

3. You should see now the new menu item "Tools/Scripts/Stablehorde". If this is not the case, something went wrong.

## Generate images
Now we are ready for generating images.

1. Start Krita and create a new document with a size between 384x384 and 1024x1024, color mode "RGB/Alpha" (default), color depth 8bit (default) and a paint layer. The generated image will have the size of the document or is a bit smaller. Check below for an explanation.
   - Stable diffusion only generates image sizes which are a multiple of 64. This means, if your document has a size of 650x512, the generated image will have a size of 640x512.
   - The larger the image, the longer you have to wait for generation. The reason is, that all servers in the cluster support 512x512, but not all larger sizes.

2. Select the new "Tools/Scripts/Stablehorde" menu item. A dialog will open, where you can enter the details for the image generation.

**Basic Tab**

   - **Generation Mode:** 
      - **Text -> Image:** Generate an image based on your prompt.
      - **Image -> Image:** Generate an image based on an init image and on your prompt.
      - **Inpainting:** Erase a part of an image and generate a new image which has the erased part filled. The erased part is filled based on your prompt. To erase, select in Krita the eraser mode or use an eraser brush preset. If you make a selection, please make sure to remove this selection before generating by clicking on the image. 

   - **NSFW:** If you want to send a prompt, which is excplicitly NSFW (Not Safe For Work). 
      - If you flag your request as NSFW, only servers, which accept NSFW prompts, work on the request. It's very likely, that it takes then longer than usual to generate the image. If you don't flag the prompt, but it is NSFW, you will receive a black image.
      - If you didn't flag your request as NSFW and don't prompt NSFW, you will receive in some cases a black image, although it's not NSFW (false positive). Just rerun the generation in that case.

   - **Seed:** This parameter is optional. If it is empty, a random seed will be generated on the server. If you use a seed, the same image is generated again in the case the same parameters for init strength, steps, etc. are used. A slightly different image will be generated, if the parameters are modified. You find the seed as part of the layer name of the layer, where the generated image is displayed. 

   - **Prompt:** How the generated image should look like.

   - **Generate:** Start image generation. The values you inserted into the dialog will be transmitted to the server, which dispatches the request now to one of the stable-diffusion servers in the cluster. Your generation request is added to the queue. You will see now the status "Queue position..." and all input elements of the dialog are disabled. Generation can take between several seconds and several minutes depending on how many other requests are already in the queue and how many stable diffusion servers joined currently the cluster.<br>When the image has been generated successfully, it will be shown as a new layer of the opened document. The used seed is shown as a part of the name of the new layer. If an error during generation occurs, the error message will be shown in the status textarea and all input elements will be enabled again.

   - **Cancel:** Close the dialog. If you cancel when you already started generation, generation will be stopped.

**Advanced Tab**
   - **Init Strength:** How much the AI should take the init image into account. The higher the value, the more will the generated image look like the init image. 0.3 is a good value to use.

   - **Prompt Strength:** How much the AI should follow the prompt. The higher the value, the more the AI will generate an image which looks like your prompt. 8 is a good value to use.

   - **Steps:** How many steps the AI should use to generate the image. The higher the value, the more the AI will work on details. But it also means, the longer the generation takes and the more the GPU is used. 50 is a good value to use.

   - **API key:** This parameter is optional. If you don't enter an API key, you run the image generation as anonymous. The downside is, that you will have then the lowest priority in the generation queue. For that reason it is recommended registering for free on [stablehorde](https://stablehorde.net) and getting an API key.

   - **Max Wait:** The maximum time in minutes you want to wait until image generation is finished. When the max time is reached, a timeout happens and the generation request is stopped.

## Troubleshooting
### Linux
- Please make sure, you have webp support for Qt5 installed, as this is the image format stablehorde uses. Execute the following command to check, if it's already supported: ```python -c "from PyQt5.Qt import *; print([bytes(x).decode('ascii') for x in QImageReader.supportedImageFormats()])"```. To install it, use ```sudo pacman -S qt5-imageformats``` (Arch) or ```sudo apt install qt5-image-formats-plugins``` (Debian).

## Limitations

   - **Stability:** Stablehorde is still pretty new and under heavy development. So, it's not unlikely, that the servers are not available for some time or unexpected errors occur.

   - **Generation speed:** Stablehorde is a cluster of stable-diffusion servers run by volunteers. The generation speed depends on how many servers are in the cluster, which hardware they use and how many others want to generate with stablehorde. The upside is, that stablehorde is free to use, the downside that the generation speed is unpredictable.

   - **Privacy:** The privacy stablehorde offers is similar to generating in a public discord channel. So, please assume, that neither your prompts nor your generated images are private.
   
   - **Features:** Currently text2img, img2img and inpainting are supported. As soon as stablehorde supports outpainting, this will be available in the plugin too.

## FAQ
**Why is the generated image smaller than the document?** Stable-diffusion only generates image sizes which are a multiple of 64. This means, if your document has a size of 650x512, the generated image will have a size of 640x512.

**Will outpainting be supported?** Pretty likely outpainting will be supported. This depends on which features the stablehorde cluster supports.

**How do I report an error or request a new feature?** Please open a new issue in this repository.


HISTORY.md
# History
## Krita Plugin
### 1.3.4
Changes
- The generated images are now transferred via Cloudflare r2.

### 1.3.3
Changes
- The new parameter r2 is transferred with value False. This disables for now the new Cloudflare r2 image download.

### 1.3.2
Changes
- Now detailed error messages are displayed. Before only the standard HTTP error messages.

### 1.3.1
Changes
- In the case no worker is available for image generation, now a message is displayed.
- Minimum size has been reduced from 512x512 to 384x384.

Bugfixes
- In the dialog init strength selector was not disabled in mode inpainting.
- In the dialog in mode inpainting was not checked if a layer with an inpainting image exists.

### 1.3.0
Changes
- Inpainting is supported now.

### 1.2.0
Changes
- Now images with sizes between 512x512 and 1024x1024 can be generated. Before only 512x512.
- The dialog has now two tabs to make the layout cleaner.
- It is now checked, if support for webp in Qt5 exists. On some Linux distributions the support is missing and needs to be installed manually.

### 1.1.1
Bugfixes
- When using img2img in some cases an error 400 occurred.
- It was not checked, if a prompt was entered. If it was missing, an error 400 occurred.

### 1.1.0
Changes
- img2img is now supported.

Bugfixes
- If the dialog was closed via the cross, the generation was not stopped.

### 1.0.1
Changes
- Now status information is displayed after generation start.

### 1.0.0
- Initial version.


krita_stablehorde/krita_stablehorde.py
# v1.3.4

import base64
import json
import ssl
import threading
import urllib
import math
import re

from krita import *

VERSION = 134

class Stablehorde(Extension):
   def __init__(self, parent):
      super().__init__(parent)

   def setup(self):
      pass

   def createActions(self, window):
      action = window.createAction("generate", "Stablehorde", "tools/scripts")
      action.triggered.connect(self.openDialog)

   def openDialog(self):
      dialog = Dialog()
      dialog.exec()

class Dialog(QDialog):
   def __init__(self):
      super().__init__(None)

      settings = self.readSettings()

      self.setWindowTitle("Stablehorde")
      self.layout = QVBoxLayout()

      # Basic Tab
      tabBasic = QWidget()
      layout = QFormLayout()

      # Generation Mode
      box = QGroupBox()
      self.modeText2Img = QRadioButton("Text -> Image")
      self.modeImg2Img = QRadioButton("Image -> Image")
      self.modeInpainting = QRadioButton("Inpainting")
      layoutV = QVBoxLayout()
      layoutV.addWidget(self.modeText2Img)
      layoutV.addWidget(self.modeImg2Img)
      layoutV.addWidget(self.modeInpainting)
      box.setLayout(layoutV)
      label = QLabel("Generation Mode")
      label.setStyleSheet("QLabel{margin-top:12px;}")
      layout.addRow(label, box)

      group = QButtonGroup()
      group.addButton(self.modeText2Img, worker.MODE_TEXT2IMG)
      group.addButton(self.modeImg2Img, worker.MODE_IMG2IMG)
      group.addButton(self.modeInpainting, worker.MODE_INPAINTING)
      group.button(settings["generationMode"]).setChecked(True)
      self.generationMode = group
      self.generationMode.buttonClicked.connect(self.handleModeChanged)

      mode = self.generationMode.checkedId()

      # NSFW
      self.nsfw = QCheckBox()
      self.nsfw.setCheckState(settings["nsfw"])
      layout.addRow("NSFW",self.nsfw)

      # Seed
      self.seed = QLineEdit()
      self.seed.setText(settings["seed"])
      layout.addRow("Seed (optional)", self.seed)

      # Prompt
      self.prompt = QTextEdit()
      self.prompt.setText(settings["prompt"])
      layout.addRow("Prompt", self.prompt)

      # Status
      self.statusDisplay = QTextEdit()
      self.statusDisplay.setReadOnly(True)
      layout.addRow("Status", self.statusDisplay)

      # Generate
      self.generateButton = QPushButton("Generate")
      self.generateButton.clicked.connect(self.generate)
      layout.addWidget(self.generateButton)

      # Space
      layoutH = QHBoxLayout()
      layoutH.addSpacing(50)
      container = QWidget()
      container.setLayout(layoutH)
      layout.addWidget(container)

      # Cancel
      cancelButton = QPushButton("Cancel")
      cancelButton.setFixedWidth(100)
      cancelButton.clicked.connect(self.reject)
      layout.addWidget(cancelButton)
      layout.setAlignment(cancelButton, Qt.AlignRight)

      tabBasic.setLayout(layout)

      # Advanced Tab
      tabAdvanced = QWidget()
      layout = QFormLayout()

      # Init Strength
      slider = QSlider(Qt.Orientation.Horizontal, self)
      slider.setRange(0, 10)
      slider.setTickInterval(1)
      slider.setValue(settings["initStrength"])
      self.initStrength = slider
      labelInitStrength = QLabel(str(self.initStrength.value()/10))
      self.initStrength.valueChanged.connect(lambda: labelInitStrength.setText(str(self.initStrength.value()/10)))
      layoutH = QHBoxLayout()
      layoutH.addWidget(self.initStrength)
      layoutH.addWidget(labelInitStrength)
      container = QWidget()
      container.setLayout(layoutH)
      layout.addRow("Init Strength", container)

      if mode == worker.MODE_TEXT2IMG or mode == worker.MODE_INPAINTING:
         self.initStrength.setEnabled(False)

      # Prompt Strength
      slider = QSlider(Qt.Orientation.Horizontal, self)
      slider.setRange(0, 20)
      slider.setTickInterval(1)
      slider.setValue(settings["promptStrength"])
      self.promptStrength = slider
      labelPromptStrength = QLabel(str(self.promptStrength.value()))
      self.promptStrength.valueChanged.connect(lambda: labelPromptStrength.setText(str(self.promptStrength.value())))
      layoutH = QHBoxLayout()
      layoutH.addWidget(self.promptStrength)
      layoutH.addWidget(labelPromptStrength)
      container = QWidget()
      container.setLayout(layoutH)
      layout.addRow("Prompt Strength", container)

      # Steps
      slider = QSlider(Qt.Orientation.Horizontal, self)
      slider.setRange(10, 200)
      slider.setTickInterval(1)
      slider.setValue(settings["steps"])
      self.steps = slider
      labelSteps = QLabel(str(self.steps.value()))
      self.steps.valueChanged.connect(lambda: labelSteps.setText(str(self.steps.value())))
      layoutH = QHBoxLayout()
      layoutH.addWidget(self.steps)
      layoutH.addWidget(labelSteps)
      container = QWidget()
      container.setLayout(layoutH)
      layout.addRow("Steps", container)

      # API Key
      self.apikey = QLineEdit()
      self.apikey.setText(settings["apikey"])
      layout.addRow("API Key (optional)", self.apikey)

      # Max Wait
      slider = QSlider(Qt.Orientation.Horizontal, self)
      slider.setRange(1, 5)
      slider.setTickInterval(1)
      slider.setValue(settings["maxWait"])
      self.maxWait = slider
      labelMaxWait = QLabel(str(self.maxWait.value()))
      self.maxWait.valueChanged.connect(lambda: labelMaxWait.setText(str(self.maxWait.value())))
      layoutH = QHBoxLayout()
      layoutH.addWidget(self.maxWait)
      layoutH.addWidget(labelMaxWait)
      container = QWidget()
      container.setLayout(layoutH)
      layout.addRow("Max Wait (minutes)", container)

      tabAdvanced.setLayout(layout)

      tabs = QTabWidget()
      tabs.addTab(tabBasic, "Basic")
      tabs.addTab(tabAdvanced, "Advanced")
      self.layout.addWidget(tabs)

      self.setLayout(self.layout)
      self.resize(350, 300)

      webpSupport = utils.checkWebpSupport()

      if webpSupport is False:
         self.generateButton.setEnabled(False)
         self.statusDisplay.setText("Your operating system doesn't support the webp image format. Please check troubleshooting section of readme on GitHub for solution.")

      update = utils.checkUpdate()

      if update["update"] is True:
         self.statusDisplay.setText(update["message"])

   def handleModeChanged(self):
      mode = self.generationMode.checkedId()

      if mode == worker.MODE_TEXT2IMG or mode == worker.MODE_INPAINTING:
         self.initStrength.setEnabled(False)
      elif mode == worker.MODE_IMG2IMG:
         self.initStrength.setEnabled(True)

   def generate(self):
      mode = self.generationMode.checkedId()
      doc = Application.activeDocument()

      # no document
      if doc is None:
         utils.errorMessage("Please open a document. Please check details.", "For image generation a document with a size between 384x384 and 1024x1024, color model 'RGB/Alpha', color depth '8-bit integer' and a paint layer is needed.")
         return
      # document has invalid color model or depth
      elif doc.colorModel() != "RGBA" or doc.colorDepth() != "U8":
         utils.errorMessage("Invalid document properties. Please check details.", "For image generation a document with color model 'RGB/Alpha', color depth '8-bit integer' is needed.")
         return
      # document too small or large
      elif doc.width() < 384 or doc.width() > 1024 or doc.height() < 384 or doc.height() > 1024:
         utils.errorMessage("Invalid document size. Please check details.", "Document needs to be between 384x384 and 1024x1024.")
         return
      # img2img/inpainting: missing init image layer
      elif (mode == worker.MODE_IMG2IMG or mode == worker.MODE_INPAINTING) and worker.getInitNode() is None:
         utils.errorMessage("Please add a visible layer which shows the init/inpainting image.", "")
         return
      # img2img/inpainting: selection has to be removed otherwise crashes krita when creating init image
      elif (mode == worker.MODE_IMG2IMG or mode == worker.MODE_INPAINTING) and doc.selection() is not None:
         utils.errorMessage("Please remove the selection by clicking on the image.", "")
         return
      # no prompt
      elif len(self.prompt.toPlainText()) == 0:
         utils.errorMessage("Please enter a prompt.", "")
         return
      else:
         self.writeSettings()
         self.setEnabledStatus(False)
         self.statusDisplay.setText("Waiting for generated image...")
         worker.generate(self)

   #override
   def customEvent(self, ev):
      if ev.type() == worker.eventId:
         if ev.updateType == UpdateEvent.TYPE_CHECKED:
            self.statusDisplay.setText(ev.message)
         elif ev.updateType == UpdateEvent.TYPE_INFO:
            self.statusDisplay.setText(ev.message)
            self.setEnabledStatus(True)
         elif ev.updateType == UpdateEvent.TYPE_ERROR:
            self.statusDisplay.setText("An error occurred: " + ev.message)
            self.setEnabledStatus(True)
         elif ev.updateType == UpdateEvent.TYPE_FINISHED:
            self.close()

   #override
   def reject(self):
      worker.cancel()
      self.writeSettings()
      super().reject()

   def readSettings(self):
      defaults = {
         "generationMode": worker.MODE_TEXT2IMG,
         "initStrength": 3,
         "prompt": "",
         "promptStrength": 8,
         "steps": 50,
         "seed": "",
         "nsfw": 0,
         "apikey": "",
         "maxWait": 5
      }

      try:
         settings = Application.readSetting("Stablehorde", "Config", None)

         if not settings:
            settings = defaults
         else:
            settings = json.loads(settings)
            missing = False

            for key in defaults:
               if not key in settings:
                  missing = True
                  break

            if missing is True:
               settings = defaults
      except Exception as ex:
         settings = defaults

      return settings

   def writeSettings(self):
      settings = {
         "generationMode": self.generationMode.checkedId(),
         "initStrength": self.initStrength.value(),
         "prompt": self.prompt.toPlainText(),
         "promptStrength": self.promptStrength.value(),
         "steps": int(self.steps.value()),
         "seed": self.seed.text(),
         "nsfw": self.nsfw.checkState(),
         "apikey": self.apikey.text(),
         "maxWait": self.maxWait.value()
      }

      try:
         settings = json.dumps(settings)
         Application.writeSetting("Stablehorde", "Config", settings)
      except Exception as ex:
         ex = ex

   def setEnabledStatus(self, status):
      self.modeText2Img.setEnabled(status)
      self.modeImg2Img.setEnabled(status)
      self.modeInpainting.setEnabled(status)

      if self.generationMode.checkedId() == worker.MODE_IMG2IMG:
         self.initStrength.setEnabled(status)

      self.promptStrength.setEnabled(status)
      self.steps.setEnabled(status)
      self.seed.setEnabled(status)
      self.nsfw.setEnabled(status)
      self.prompt.setEnabled(status)
      self.apikey.setEnabled(status)
      self.maxWait.setEnabled(status)
      self.generateButton.setEnabled(status)

class Worker():
   API_ROOT = "https://stablehorde.net/api/v2/"
   CHECK_WAIT = 5
   MODE_TEXT2IMG = 1
   MODE_IMG2IMG = 2
   MODE_INPAINTING = 3

   dialog = None
   checkMax = None
   checkCounter = 0
   id = None
   cancelled = False

   eventId = QEvent.registerEventType()

   ssl._create_default_https_context = ssl._create_unverified_context

   def getInitImage(self):
      doc = Application.activeDocument()
      nodeInit = self.getInitNode()

      if nodeInit is not None:
         if doc.selection() is not None:
            raise Exception("Selection has to be removed before creating init image.")

         bytes = nodeInit.pixelData(0, 0, doc.width(), doc.height())
         image = QImage(bytes.data(), doc.width(), doc.height(), QImage.Format_RGBA8888).rgbSwapped()
         bytes = QByteArray()
         buffer = QBuffer(bytes)
         image.save(buffer, "WEBP")
         data = base64.b64encode(bytes.data())
         data = data.decode("ascii")
         return data
      else:
         raise Exception("No layer with init image found.")

   def getInitNode(self):
      doc = Application.activeDocument()
      nodes = doc.topLevelNodes()

      nodeInit = None

      for node in nodes:
         if node.visible() is True:
            nodeInit = node

      return nodeInit

   def displayGenerated(self, images):
      for image in images:
         seed = image["seed"]

         if re.match("^https.*", image["img"]):
            response = urllib.request.urlopen(image["img"])
            bytes = response.read()
         else:
            bytes = base64.b64decode(image["img"])
            bytes = QByteArray(bytes)

         image = QImage()
         image.loadFromData(bytes, 'WEBP')
         ptr = image.bits()
         ptr.setsize(image.byteCount())

         doc = Application.activeDocument()
         root = doc.rootNode()
         node = doc.createNode("Stablehorde " + str(seed), "paintLayer")
         root.addChildNode(node, None)
         node.setPixelData(QByteArray(ptr.asstring()), 0, 0, image.width(), image.height())
         doc.waitForDone()
         doc.refreshProjection()

   def getImages(self):
      url = self.API_ROOT + "generate/status/" + self.id
      response = urllib.request.urlopen(url)
      data = response.read()
      data = json.loads(data)
      return data["generations"]

   def checkStatus(self):
      try:
         url = self.API_ROOT + "generate/check/" + self.id
         response = urllib.request.urlopen(url)
         data = response.read()
         data = json.loads(data)

         self.checkCounter = self.checkCounter + 1

         if self.checkCounter < self.checkMax and data["done"] is False and self.cancelled is False:
            if data["is_possible"] is True:
               if data["processing"] == 0:
                  message = "Queue position: " + str(data["queue_position"]) + ", Wait time: " + str(data["wait_time"]) + "s"
               elif data["processing"] > 0:
                  message = "Generating..."

               ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_CHECKED, message)
               QApplication.postEvent(self.dialog, ev)

               timer = threading.Timer(self.CHECK_WAIT, self.checkStatus)
               timer.start()
            else:
               self.cancelled = True
               message = "Currently no worker available to generate your image. Please try again later."
               ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_INFO, message)
               QApplication.postEvent(self.dialog, ev)
         elif self.checkCounter == self.checkMax and self.cancelled == False:
            self.cancelled = True
            minutes = (self.checkMax * self.CHECK_WAIT)/60
            message = "Image generation timed out after " + str(minutes) + " minutes. Please try it again later."
            ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_INFO, message)
            QApplication.postEvent(self.dialog, ev)
         elif data["done"] == True and self.cancelled == False:
            images = self.getImages()
            self.displayGenerated(images)

            ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_FINISHED)
            QApplication.postEvent(self.dialog, ev)

         return
      except urllib.error.HTTPError as ex:
         try:
            data = ex.read()
            data = json.loads(data)

            if "message" in data:
               message = data["message"]
            else:
               message = str(ex)
         except Exception:
            message = str(ex)

         ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_ERROR, message)
         QApplication.postEvent(self.dialog, ev)
      except Exception as ex:
         ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_ERROR, str(ex))
         QApplication.postEvent(self.dialog, ev)

   def generate(self, dialog):
      self.dialog = dialog
      self.checkCounter = 0
      self.cancelled = False
      self.id = None
      self.checkMax = (self.dialog.maxWait.value() * 60)/self.CHECK_WAIT

      try:
         nsfw = True if self.dialog.nsfw.isChecked() else False

         params = {
            "cfg_scale": self.dialog.promptStrength.value(),
            "steps": int(self.dialog.steps.value()),
            "seed": self.dialog.seed.text()
         }

         data = {
            "prompt": self.dialog.prompt.toPlainText(),
            "params": params,
            "nsfw": nsfw,
            "censor_nsfw": False,
            "r2": True
         }

         doc = Application.activeDocument()

         if doc.width() % 64 != 0:
            width = math.floor(doc.width()/64) * 64
         else:
            width = doc.width()

         if doc.height() % 64 != 0:
            height = math.floor(doc.height()/64) * 64
         else:
            height = doc.height()

         params.update({"width": width})
         params.update({"height": height})

         mode = self.dialog.generationMode.checkedId()

         if mode == worker.MODE_IMG2IMG:
            init = self.getInitImage()
            data.update({"source_image": init})
            data.update({"source_processing": "img2img"})
            params.update({"denoising_strength": round((1 - self.dialog.initStrength.value()/10), 1)})
         elif mode == worker.MODE_INPAINTING:
            init = self.getInitImage()
            models = ["stable_diffusion_inpainting"]
            data.update({"source_image": init})
            data.update({"source_processing": "inpainting"})
            data.update({"models": models})

         data = json.dumps(data).encode("utf-8")

         apikey = "0000000000" if self.dialog.apikey.text() == "" else self.dialog.apikey.text()
         headers = {"Content-Type": "application/json", "Accept": "application/json", "apikey": apikey}

         url = self.API_ROOT + "generate/async"

         request = urllib.request.Request(url=url, data=data, headers=headers)
         self.dialog.statusDisplay.setText("Waiting for generated image...")

         response = urllib.request.urlopen(request)
         data = response.read()

         try:
            data = json.loads(data)
            self.id = data["id"]
         except Exception as ex:
            raise Exception(data)

         self.checkStatus()
      except urllib.error.HTTPError as ex:
         try:
            data = ex.read()
            data = json.loads(data)

            if "message" in data:
               message = data["message"]
            else:
               message = str(ex)
         except Exception:
            message = str(ex)

         ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_ERROR, message)
         QApplication.postEvent(self.dialog, ev)
      except Exception as ex:
         ev = UpdateEvent(worker.eventId, UpdateEvent.TYPE_ERROR, str(ex))
         QApplication.postEvent(self.dialog, ev)

      return

   def cancel(self):
      self.cancelled = True

class Utils():
   updateChecked = False

   def errorMessage(self, text, detailed):
      msgBox = QMessageBox()
      msgBox.setWindowTitle("Stablehorde")
      msgBox.setText(text)
      msgBox.setDetailedText(detailed)
      msgBox.setStyleSheet("QLabel{min-width: 300px;}")
      msgBox.exec()

   def checkUpdate(self):
      if self.updateChecked is False:
         try:
            url = "https://raw.githubusercontent.com/blueturtleai/krita-stable-diffusion/main/version.json"
            response = urllib.request.urlopen(url)
            data = response.read()
            data = json.loads(data)

            self.updateChecked = True

            if VERSION < int(data["version"]):
               return {"update": True, "message": data["message"]}
            else:
               return {"update": False}
         except Exception as ex:
            return {"update": False}
      else:
         return {"update": False}

   def checkWebpSupport(self):
      formats = QImageReader.supportedImageFormats()
      found = False

      for format in formats:
         if format.data().decode("ascii").lower() == "webp":
            found = True
            break

      return found

class UpdateEvent(QEvent):
   TYPE_CHECKED = 0
   TYPE_ERROR = 1
   TYPE_INFO = 2
   TYPE_FINISHED = 3

   def __init__(self, eventId, updateType, message = ""):
      self.updateType = updateType
      self.message = message
      super().__init__(eventId)

Krita.instance().addExtension(Stablehorde(Krita.instance()))
utils = Utils()
worker = Worker()
#dialog = Dialog()
#dialog.exec()
